src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -230,8 +242,11 @@ code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
     ordered_snapshot_symbols.begin(),
     ordered_snapshot_symbols.end(),
     [this](const snapshot_pairt &left, const snapshot_pairt &right) {
-      return pointer_depth(left.second.symbol_expr().type()) <
-             pointer_depth(right.second.symbol_expr().type());
+      if(refers_to(right.second.value, left.first))
```

Comment:
For example, say `int* p = malloc(8); int *q = p+1;` is the source code for which we want to take the memory snapshot. Then the snapshot will contain: `int tmp[2]; int* p = tmp[0]; int* q = tmp[1];` And both `p` and `q` `refer_to` `tmp`. I sort the symbols based on this notion of _reference_ so that when building the harness entry function, `tmp` is assigned before the two pointers.